### PR TITLE
[CSM-377] Fix "Not enough money" bug

### PR DESCRIPTION
**Pull request by flyingleafe**
This fixes the bug by:
1) Fixing `localChainDifficulty`, which stopped representing actual blockchain syncing progress after last changes in blockchain recovery
2) Making sure that `BListener`s are not applied to wallets when it's inappropriate to do so.

### CODE REVIEW

**Comment by gromakovsky** [comment]
What is `cardano-sl-0.5-staging`? According to #1147 this branch will never be merged. Isn't it better to make this PR to `cardano-sl-0.5` and then rebase staging branch?

File: src/Pos/Wallet/Web/BListener.hs [hunk 1]
```diff
@@ -50,6 +50,27 @@ instance ( MonadRealDB m
     onApplyBlocks = onApplyTracking
     onRollbackBlocks = onRollbackTracking
 
+walletGuard
```
Comment (pva701) [inline src/Pos/Wallet/Web/BListener.hs]: nice code!

File: src/Pos/Wallet/Web/BListener.hs [hunk 2]
```diff
@@ -50,6 +50,27 @@ instance ( MonadRealDB m
...
+       , WithLogger m
+       )
+    => HeaderHash
+    -> CId Wal
+    -> m ()
+    -> m ()
+walletGuard curTip wAddr action = WS.getWalletSyncTip wAddr >>= \case
+    Nothing ->
+        logWarning $
+        sformat ("Wallet Tracking: there is no syncTip corresponding to wallet #"%build) wAddr
```
Comment (pva701) [inline src/Pos/Wallet/Web/BListener.hs]: btw, this message is printed on start for `myRootAddress` (or something like that).
there are some wallet-addresses without `walletSyncTip` it was unexpected for me

Comment (pva701) [inline src/Pos/Wallet/Web/BListener.hs]: so we will see some warning on start

**Comment by domenkozar** [comment]
Pushed a fix from `master` to address the macOS build failure.

**Comment by domenkozar** [comment]
@flyingleafe @gromakovsky @fersel this is ready to be merged then.

**Comment by gromakovsky** [comment]
@flyingleafe don't forget to merge to master.

**Comment by gromakovsky** [comment]
@flyingleafe ping.

**Comment by gromakovsky** [comment]
@flyingleafe daily ping.

**Comment by flyingleafe** [comment]
@gromakovsky https://github.com/input-output-hk/cardano-sl/pull/1225

**Pull request merged**